File: android/core/designsystem/src/main/kotlin/mv/muraka/core/designsystem/component/Readouts.kt
package mv.muraka.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import mv.muraka.core.designsystem.theme.MurakaTheme
import mv.muraka.core.designsystem.theme.ReadoutStyle
import mv.muraka.core.model.Condition
import mv.muraka.core.model.Prediction
import mv.muraka.core.model.SightingDisplayStatus
import kotlin.math.roundToInt

/**
 * A measured quantity, monospaced with tabular figures.
 *
 * Every number in this app goes through here: coordinates, depths, severities, counts,
 * model versions. It is the cheapest piece of family resemblance the three clients have
 * and the strongest — columns line up, and the interface reads like an instrument.
 *
 * Only the family is fixed. The size comes from [style], so it still scales with the
 * user's font-size setting.
 */
@Composable
fun Readout(
    value: String,
    modifier: Modifier = Modifier,
    label: String? = null,
    style: androidx.compose.ui.text.TextStyle = MaterialTheme.typography.bodyMedium,
) {
    Column(modifier = modifier) {
        label?.let {
            Text(
                text = it,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Text(
            text = value,
            style = style.merge(ReadoutStyle),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

/**
 * What the model made of a photograph.
 *
 * **Severity leads, not the label.** "62% bleached" tells a contributor something that
 * "bleached" does not — it is the difference between a reef in trouble and a reef with a
 * few pale colonies, and it is the number researchers actually work with.
 *
 * The model version is shown because it is provenance: `fake-0.0.0` means no trained
 * model is loaded yet, and a reader of the project needs to be able to tell which
 * screenshots predate the real one.
 */
@Composable
fun PredictionReadout(prediction: Prediction, verified: Boolean, modifier: Modifier = Modifier) {
    val extent = (prediction.severity * 100).roundToInt()
    val confidence = (prediction.confidence * 100).roundToInt()

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            SeveritySwatch(prediction.severity)
            Text(
                text = "$extent% bleached",
                style = MaterialTheme.typography.titleMedium.merge(ReadoutStyle),
            )
            ProvenanceChip(verified = verified)
        }

        Row(horizontalArrangement = Arrangement.spacedBy(20.dp)) {
            Readout(label = "Confidence", value = "$confidence%")
            Readout(label = "Grid", value = "${prediction.patchGrid}×${prediction.patchGrid}")
            prediction.inferenceMs?.let { Readout(label = "Inference", value = "$it ms") }
        }

        Readout(
            label = "Model",
            value = prediction.modelVersion,
            style = MaterialTheme.typography.bodySmall,
        )
    }
}

/**
 * A block of the severity ramp.
 *
 * A graphic, never text: the condition colours are fills, and body text set in them would
 * fail contrast against half the surfaces it could land on. The number beside it carries
 * the meaning.
 */
@Composable
fun SeveritySwatch(severity: Double, modifier: Modifier = Modifier) {
    val reef = MurakaTheme.reef
    Surface(
        modifier = modifier
            .clearAndSetSemantics { },
        shape = RoundedCornerShape(4.dp),
        color = reef.severity(severity),
    ) {
        androidx.compose.foundation.layout.Box(Modifier.padding(horizontal = 8.dp, vertical = 8.dp))
    }
}

/**
 * The contributor-facing status of a sighting.
 *
 * The vocabulary is [SightingDisplayStatus] and nothing else — there is no "Synced",
 * because a local flag claiming the upload worked is a claim rather than a fact. The
 * colour is decoration; the word is the information, which is what makes this legible in
 * a greyscale screenshot.
 */
@Composable
fun StatusPill(status: SightingDisplayStatus, modifier: Modifier = Modifier) {
    val reef = MurakaTheme.reef
    val colour = when (status) {
        SightingDisplayStatus.VERIFIED_BY_EXPERT -> reef.verified
        SightingDisplayStatus.NOT_USABLE -> reef.rust
        SightingDisplayStatus.FAILED -> reef.rust
        SightingDisplayStatus.WAITING_TO_UPLOAD,
        SightingDisplayStatus.UPLOADING,
        SightingDisplayStatus.CHECKING,
        SightingDisplayStatus.PHOTOS_PENDING,
        -> reef.amber

        SightingDisplayStatus.ANALYSING,
        SightingDisplayStatus.AWAITING_REVIEW,
        -> MaterialTheme.colorScheme.onSurfaceVariant
    }

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(6.dp),
        color = colour.copy(alpha = 0.14f),
        contentColor = colour,
    ) {
        Text(
            text = status.label,
            style = MaterialTheme.typography.labelMedium,
            modifier = Modifier.padding(horizontal = 8.dp, vertical = 3.dp),
        )
    }
}

/** "Healthy" or "Bleached", as a word, for places where only the label is available. */
@Composable
fun ConditionLabel(condition: Condition?, modifier: Modifier = Modifier) {
    Text(
        modifier = modifier,
        text = when (condition) {
            Condition.HEALTHY -> "Healthy"
            Condition.BLEACHED -> "Bleached"
            null -> "Not yet assessed"
        },
        style = MaterialTheme.typography.bodyMedium,
    )
}

File: ios/Muraka/Core/DesignSystem/Readouts.swift
import UIKit

/// A measured quantity, monospaced with tabular figures.
///
/// Every number in this app goes through here: coordinates, depths, severities, counts, model
/// versions. It is the cheapest piece of family resemblance the three clients have and the
/// strongest - columns line up, and the interface reads like an instrument.
///
/// The font is `monospacedSystemFont` at the size the text style asks for, so it still scales
/// with Dynamic Type. A fixed point size here would be a bug, not a design.
final class ReadoutView: UIStackView {
    private let captionLabel = UILabel()
    private let valueLabel = UILabel()

    init(caption: String? = nil, value: String, style: UIFont.TextStyle = .body) {
        super.init(frame: .zero)
        axis = .vertical
        spacing = 2

        if let caption {
            captionLabel.text = caption
            captionLabel.font = .preferredFont(forTextStyle: .caption2)
            captionLabel.adjustsFontForContentSizeCategory = true
            captionLabel.textColor = .secondaryLabel
            addArrangedSubview(captionLabel)
        }

        valueLabel.text = value
        valueLabel.font = Self.monospaced(style)
        valueLabel.adjustsFontForContentSizeCategory = true
        valueLabel.textColor = .label
        addArrangedSubview(valueLabel)
    }

    @available(*, unavailable)
    required init(coder _: NSCoder) {
        // UIStackView's init(coder:) is non-failable, so this cannot return nil the way a
        // UIView subclass's can. Nothing in this app comes from a XIB, so reaching it is a
        // programmer error rather than a runtime condition.
        fatalError("Muraka builds its views in code; init(coder:) is unused")
    }

    var value: String? {
        get { valueLabel.text }
        set { valueLabel.text = newValue }
    }

    /// A monospaced font that still tracks Dynamic Type.
    static func monospaced(_ style: UIFont.TextStyle, weight: UIFont.Weight = .medium) -> UIFont {
        let descriptor = UIFontDescriptor.preferredFontDescriptor(withTextStyle: style)
        let base = UIFont.monospacedSystemFont(ofSize: descriptor.pointSize, weight: weight)
        return UIFontMetrics(forTextStyle: style).scaledFont(for: base)
    }
}

/// The contributor-facing status of a sighting.
///
/// The vocabulary is ``SightingDisplayStatus`` and nothing else - there is no "Synced",
/// because a local flag claiming the upload worked is a claim rather than a fact. The colour
/// is decoration; the **word** is the information, which is what makes this legible in a
/// greyscale screenshot.
final class StatusPillView: UIView {
    private let label = UILabel()

    init(status: SightingDisplayStatus) {
        super.init(frame: .zero)

        let colour: UIColor = switch status {
        case .verifiedByExpert: ReefPalette.verified
        case .notUsable, .failed: ReefPalette.rust
        case .waitingToUpload, .uploading, .checking, .photosPending: ReefPalette.amber
        case .analysing, .awaitingReview: .secondaryLabel
        }

        backgroundColor = colour.withAlphaComponent(0.14)
        layer.cornerRadius = 6
        layer.cornerCurve = .continuous

        label.translatesAutoresizingMaskIntoConstraints = false
        label.text = status.label
        label.font = .preferredFont(forTextStyle: .caption1)
        label.adjustsFontForContentSizeCategory = true
        label.textColor = colour
        label.numberOfLines = 1
        addSubview(label)

        NSLayoutConstraint.activate([
            label.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 8),
            label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -8),
            label.topAnchor.constraint(equalTo: topAnchor, constant: 3),
            label.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -3),
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { nil }
}

/// A block of the severity ramp.
///
/// A graphic, never text: the condition colours are fills, and body text set in them would
/// fail contrast against half the surfaces it could land on. The number beside it carries the
/// meaning.
final class SeveritySwatchView: UIView {
    init(severity: Double) {
        super.init(frame: .zero)
        backgroundColor = ReefPalette.severity(severity)
        layer.cornerRadius = 4
        layer.cornerCurve = .continuous
        isAccessibilityElement = true
        accessibilityLabel = "Severity \(Int((severity * 100).rounded())) percent"
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { nil }

    override var intrinsicContentSize: CGSize { CGSize(width: 18, height: 18) }
}

/// "just now", "12 minutes ago", "3 days ago".
///
/// Used for two different things, and the difference matters: the age of a *sighting*, and
/// the age of what the app *knows* about it. The second is what lets the interface say
/// "checked 20 minutes ago" instead of presenting a stale status as current.
enum RelativeTime {
    static func describe(_ date: Date, now: Date = Date()) -> String {
        let elapsed = now.timeIntervalSince(date)
        // A clock reading slightly ahead should not produce "in 3 seconds".
        guard elapsed > 60 else { return "just now" }

        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .full
        return formatter.localizedString(for: date, relativeTo: now)
    }
}
